src/aeso_ast_infer_types.erl

```diff
@@ -2309,16 +2426,29 @@ unify1(_Env, {bytes_t, _, Len}, {bytes_t, _, Len}, _When) ->
 unify1(Env, {if_t, _, {id, _, Id}, Then1, Else1}, {if_t, _, {id, _, Id}, Then2, Else2}, When) ->
     unify(Env, Then1, Then2, When) andalso
     unify(Env, Else1, Else2, When);
+
+unify1(_Env, {fun_t, _, _, _, _}, {fun_t, _, _, var_args, _}, When) ->
+    type_error({unify_varargs, When});
+unify1(_Env, {fun_t, _, _, var_args, _}, {fun_t, _, _, _, _}, When) ->
+    error({unify_varargs, When});
 unify1(Env, {fun_t, _, Named1, Args1, Result1}, {fun_t, _, Named2, Args2, Result2}, When)
-    when length(Args1) == length(Args2) ->
-    unify(Env, Named1, Named2, When) andalso
+  when length(Args1) == length(Args2) ->
+    {Named1_, Named2_} =
+        if is_list(Named1) andalso is_list(Named2) ->
+                {lists:keysort(3, Named1), lists:keysort(3, Named2)};
```

Comment:
This seems wrong. The compiler (`ast_to_fcode`) will put the arguments in the order they are listed in the type, so two functions with named arguments listed in different orders will get compiled differently.